lambda/translcore.ml

```diff
@@ -460,99 +460,169 @@ and transl_exp0 ~in_new_scope ~scopes e =
       Lwhile(transl_exp ~scopes cond,
              event_before ~scopes body (transl_exp ~scopes body))
   
-  | Texp_arr_comprehension (body, type_comps) -> 
-    let make_array size init= 
-      let prim_make_arr = 
+  | Texp_arr_comprehension (body, blocks) -> 
+    (*One block consists of comprehension statements connected by "and".*)
+    (* TODO: Move some of the transl methods out? Which ones shoudl 
+       I move out? All of them?*)
+    (*TODO: Replace Loc_unknown for errors that can occur because of user
+      error with something usefull*)
+    let transl_make_array size init= 
+      let transl_prim_make_arr = 
+        (*Todo: What is arity?*)
         Primitive.simple ~name:"caml_make_vect" ~arity:2 ~alloc:true 
       in
-      Lprim (Pccall prim_make_arr, [size; init], Loc_unknown)
+      Lprim (Pccall transl_prim_make_arr, [size; init], Loc_unknown)
     in
-    let l_array_set arr at v=
+    let empty_arr = 
+      Lprim( (Pmakearray(Pgenarray, Immutable)), [] ,Loc_unknown) 
+    in 
+    let transl_array_set arr at v=
       Lprim( (Parraysets(Pgenarray)), [arr; at; v] ,Loc_unknown) 
     in
-    let l_array_get arr at=
+    let transl_array_get arr at=
       Lprim( (Parrayrefs(Pgenarray)), [arr; at] ,Loc_unknown) 
     in
-    let l_int_op op a b = 
-      Lprim( op, [a; b] ,Loc_unknown) in
+    let transl_int_op op a b = Lprim( op, [a; b] ,Loc_unknown) in
     let int n = Lconst (Const_base (Const_int n)) in   
-    let construct_base_body bdy counter=
-      Lsequence( 
-        bdy,
-        Lassign(counter, l_int_op Paddint (int 1) (Lvar(counter)))) 
-    in
-    let rec create_inner_loop ~is_base type_comp ls res counter =
+    let transl_loop ~type_comp ~body =
       match type_comp with 
       | In (id, _, e2) -> 
-          let in_ = (transl_exp ~scopes e2) in
-          let len = Lprim( (Parraylength(Pgenarray)), [in_], Loc_unknown) in
-          let index = Ident.create_local "index" in    
-          if is_base then
-            let loop = 
-              Lfor(index, (int 0), (l_int_op Psubint len (int 1)), Upto, 
-              construct_base_body (
-                  l_array_set 
-                    (Lvar(res)) 
-                    (Lvar(counter)) 
-                    (Llet(Strict,Pintval, id, 
-                      (l_array_get  in_ (Lvar(index))), 
-                      (transl_exp ~scopes  body)))
-                  ) counter)
-            in 
-            loop, len
-          else 
-            let inner_body, inner_len = nested_for ls res counter in
-            let loop = 
-              Lfor(index, 
-                (int 0), (l_int_op Psubint len (int 1)), Upto, 
-                (Llet(
-                  Strict,Pintval, id, (l_array_get  in_ (Lvar(index))), 
-                  inner_body)))
-            in 
-            loop, l_int_op Pmulint inner_len len
-            
+        let in_ = (transl_exp ~scopes e2) in
+        let len = Lprim( (Parraylength(Pgenarray)), [in_], Loc_unknown) in
```

Comment:
I dont know exactly how Ocaml does that but should a length lookup be the same as a variable lookup?
So this would even add overhead but having another variable.

Reply:
The length lookup will be a read from memory, whilst reading a mutable variable will be using a register. Sometimes it is better to re-read a value from memory -- e.g. if you are going to hold the value live for a long time -- but here I expect using a variable to be better.

Reply:
Sadly, its not as easy as just defining  a new Let variable in this part of the code. Since then it will be out of scope when used the second time. 
I think transl_loops would need to return a (Ident.t * lambda) list s.t the caller can the init all the lengths such that they are in the correct scope. 
Any suggestions?

Reply:
Having `transl_loops` return `(Ident.t * lambda) list` is what I was suggesting above. I think it is probably the best way to write this code.

Reply:
Oh, yeah. Thanks!